File: src/PolyTrader.Core/Strategy/ShouldPlaceTrendBet.cs
using PolyTrader.Core.Models;

namespace PolyTrader.Core.Strategy;

public static class ShouldPlaceTrendBet
{
    public static int CountBarsSinceFlipAtOpen(IReadOnlyList<bool> bosFlipAt, int index)
    {
        var count = 0;
        for (var j = index - 1; j >= 0; j--)
        {
            if (bosFlipAt[j])
            {
                break;
            }

            count++;
        }

        return count;
    }

    public static double DistanceFromStructureAtOpen(
        IReadOnlyList<ChartCandle> candles,
        int index,
        MarketTrend trend,
        int structureLookback)
    {
        var open = candles[index].Open;
        if (trend == MarketTrend.Long)
        {
            return open - ReferenceLow(candles, index, structureLookback);
        }

        return ReferenceHigh(candles, index, structureLookback) - open;
    }

    public static bool AtOpen(
        int index,
        IReadOnlyList<ChartCandle> candles,
        IReadOnlyList<MarketTrend> trendAtOpen,
        IReadOnlyList<bool> bosFlipAt,
        TrendBetStrategyParams parameters)
    {
        var lookback = Math.Max(1, parameters.StructureLookback);
        var barsSinceFlip = CountBarsSinceFlipAtOpen(bosFlipAt, index);

        if (parameters.MinBarsSinceFlip > 0 && barsSinceFlip < parameters.MinBarsSinceFlip)
        {
            return false;
        }

        if (parameters.MaxBarsSinceFlip > 0 && barsSinceFlip > parameters.MaxBarsSinceFlip)
        {
            return false;
        }

        if (parameters.MinDistanceFromStructure > 0)
        {
            var distance = DistanceFromStructureAtOpen(
                candles,
                index,
                trendAtOpen[index],
                lookback);
            if (distance < parameters.MinDistanceFromStructure)
            {
                return false;
            }
        }

        return true;
    }

    private static double ReferenceLow(IReadOnlyList<ChartCandle> candles, int index, int lookback)
    {
        var reference = double.PositiveInfinity;
        for (var k = Math.Max(0, index - lookback); k < index; k++)
        {
            reference = Math.Min(reference, candles[k].Low);
        }

        return reference;
    }

    private static double ReferenceHigh(IReadOnlyList<ChartCandle> candles, int index, int lookback)
    {
        var reference = double.NegativeInfinity;
        for (var k = Math.Max(0, index - lookback); k < index; k++)
        {
            reference = Math.Max(reference, candles[k].High);
        }

        return reference;
    }
}

File: client/src/utils/chart/shouldPlaceTrendBet.ts
import type { ChartCandle } from '@/types/candle';
import type { TrendBetStrategyParams } from '@/types/trendBetStrategy';
import type { MarketTrend } from '@/utils/chart/detectBreakOfStructure';

export function countBarsSinceFlipAtOpen(
	bosFlipAt: boolean[],
	index: number,
): number {
	let count = 0;
	for (let j = index - 1; j >= 0; j--) {
		if (bosFlipAt[j]) {
			break;
		}
		count++;
	}
	return count;
}

function referenceLow(
	candles: ChartCandle[],
	index: number,
	lookback: number,
): number {
	let ref = Infinity;
	for (let k = Math.max(0, index - lookback); k < index; k++) {
		ref = Math.min(ref, candles[k].low);
	}
	return ref;
}

function referenceHigh(
	candles: ChartCandle[],
	index: number,
	lookback: number,
): number {
	let ref = -Infinity;
	for (let k = Math.max(0, index - lookback); k < index; k++) {
		ref = Math.max(ref, candles[k].high);
	}
	return ref;
}

/** Distance from open to structure break level (positive = safe side). */
export function distanceFromStructureAtOpen(
	candles: ChartCandle[],
	index: number,
	trend: MarketTrend,
	structureLookback: number,
): number {
	const open = candles[index].open;
	if (trend === 'long') {
		return open - referenceLow(candles, index, structureLookback);
	}
	return referenceHigh(candles, index, structureLookback) - open;
}

/**
 * Whether to open a bet at the start of bar `index`.
 * Uses only prior-bar BoS flags, structure levels, and this bar's open.
 */
export function shouldPlaceTrendBetAtOpen(
	index: number,
	candles: ChartCandle[],
	trendAtOpen: MarketTrend[],
	bosFlipAt: boolean[],
	params: TrendBetStrategyParams,
): boolean {
	const lookback = Math.max(1, Math.floor(params.structureLookback));
	const barsSinceFlip = countBarsSinceFlipAtOpen(bosFlipAt, index);

	if (params.minBarsSinceFlip > 0 && barsSinceFlip < params.minBarsSinceFlip) {
		return false;
	}
	if (params.maxBarsSinceFlip > 0 && barsSinceFlip > params.maxBarsSinceFlip) {
		return false;
	}

	if (params.minDistanceFromStructure > 0) {
		const distance = distanceFromStructureAtOpen(
			candles,
			index,
			trendAtOpen[index],
			lookback,
		);
		if (distance < params.minDistanceFromStructure) {
			return false;
		}
	}

	return true;
}
